jps-plugin/src/org/jetbrains/kotlin/compilerRunner/KotlinCompilerRunner.kt

```diff
@@ -0,0 +1,255 @@
+            libraryFiles: List<String>,
+            outputFile: File) {
+        val arguments = mergeBeans(commonArguments, k2jsArguments)
+        setupK2JsArguments(outputFile, sourceFiles, libraryFiles, arguments)
+
+        runCompiler(K2JS_COMPILER, arguments, compilerSettings.additionalArguments, messageCollector, collector, environment)
+    }
+
+    private fun ProcessCompilerOutput(
+            messageCollector: MessageCollector,
+            collector: OutputItemsCollector,
+            stream: ByteArrayOutputStream,
+            exitCode: String) {
+        val reader = BufferedReader(StringReader(stream.toString()))
+        CompilerOutputParser.parseCompilerMessagesFromReader(messageCollector, reader, collector)
+
+        if (INTERNAL_ERROR == exitCode) {
+            reportInternalCompilerError(messageCollector)
+        }
+    }
+
+    private fun reportInternalCompilerError(messageCollector: MessageCollector) {
+        messageCollector.report(ERROR, "Compiler terminated with internal error", CompilerMessageLocation.NO_LOCATION)
+    }
+
+    private fun runCompiler(
+            compilerClassName: String,
+            arguments: CommonCompilerArguments,
+            additionalArguments: String,
+            messageCollector: MessageCollector,
+            collector: OutputItemsCollector,
+            environment: CompilerEnvironment) {
+        try {
+            messageCollector.report(INFO, "Using kotlin-home = " + environment.kotlinPaths.homePath, CompilerMessageLocation.NO_LOCATION)
+
+            val argumentsList = ArgumentUtils.convertArgumentsToStringList(arguments)
+            argumentsList.addAll(StringUtil.split(additionalArguments, " "))
+
+            val argsArray = ArrayUtil.toStringArray(argumentsList)
+
+            if (!tryCompileWithDaemon(messageCollector, collector, environment, argsArray)) {
+                // otherwise fallback to in-process
+
+                val stream = ByteArrayOutputStream()
+                val out = PrintStream(stream)
+
+                val rc = CompilerRunnerUtil.invokeExecMethod(compilerClassName, argsArray, environment, messageCollector, out)
+
+                // exec() returns an ExitCode object, class of which is loaded with a different class loader,
+                // so we take it's contents through reflection
+                ProcessCompilerOutput(messageCollector, collector, stream, getReturnCodeFromObject(rc))
+            }
+        }
+        catch (e: Throwable) {
+            MessageCollectorUtil.reportException(messageCollector, e)
+            reportInternalCompilerError(messageCollector)
+        }
+
+    }
+
+    private fun tryCompileWithDaemon(messageCollector: MessageCollector,
+                                     collector: OutputItemsCollector,
+                                     environment: CompilerEnvironment,
+                                     argsArray: Array<String>): Boolean {
+        if (isDaemonEnabled()) {
```

Comment:
I tried and I don't like how it looks: understandable structure is lost. Left it as is.